### PR TITLE
WIP: Color scheme feature

### PR DESCRIPTION
**Pull request by optizone8**
I couldn't make that 7 commits to 1 to PR, so I decided just create new branch and new PR to avoid similar problem in future.

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
reference: PR #706

File: src/widgets/ColorSchemePrefWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                                                     Qt::UserRole).value<ColorOption>().textColor);
+    QStyleOptionViewItem op = option;
+
+    op.palette = pal;
+    op.displayAlignment = Qt::AlignCenter;
+
+    if (option.state & QStyle::State_Selected) {
+        QLinearGradient lgrd = QLinearGradient(option.rect.topLeft(), option.rect.bottomRight());
+        QColor highlighted = QApplication::palette().highlight().color(),
+               highlightedOpague = highlighted;
```
Comment (xarkes) [inline src/widgets/ColorSchemePrefWidget.cpp]: Opaque, no?

**Comment by thestr4ng3r** [comment]
I like how the ui is done and the code looks good too. But it probably doesn't really make much sense to merge it until it actually works with the `eco` themes.
We should have a typical ui as you often have it in IDEs and stuff where you can duplicate and edit themes. Because the default themes are stored in `<prefix>/share/radare2/.../cons`, the themes there should not be editable or deletable. For user themes, unless r2 already has such a path, we should use https://doc.qt.io/qt-5/qstandardpaths.html#writableLocation with `QStandardPaths::AppDataLocation` and no own system-dependent path logic.

**Comment by optizone8** [comment]
Of course, it is not ready to merge. This PR was created to you can stop me doing wrong. And to make some comments, discuss.

**Comment by optizone8** [comment]
Question. When DisassemblyWidget highlights opcodes, comments etc and from where it takes rules for?

**Comment by xarkes** [comment]
Currently all colors are passed from the r2 library via `pdj` or similar. So when you change a theme, it calls `eco` to load a theme into the r2 library, which will impact the output of every `pd` command.

**Comment by optizone8** [comment]
I see, but `DisassemblyWidget` calls only `Core()->disassembleLines(...)`, that returns text without information of colors. So I can't figure out how it colorize asm.

And does a command adding/changing color scheme exist in r2 library?
Because I see 2 way to implement this feature:
1) Force `DisassemblyWidget` and `DisassemblyGraphView` to use our colors instead of r2 (not the best way, I think)
2) Change r2 color schemes

**Comment by thestr4ng3r** [comment]
Yes, option 2 is the way to go. Depending on where you installed r2, take a look at the files in `<prefix>/share/radare2/.../cons`. Those are the themes, which are basically just r2 scripts. Also take a look at the `ec?` command.

The colorizing comes with the disassembly. Normally, r2 prints ansi escape codes for colorized disassembly in the terminal. In Cutter, we enable the `scr.html` option in r2, so it converts these to html and we render that in the DisassemblyWidget (it's not a WebView, but accepts a subset of HTML).

**Comment by optizone8** [comment]
There is some little trouble. I added option `copy color scheme`, so you can just copy one and then edit it. But standard scripts (`<prefix>/share/radare2/.../cons`) contain not all instructions (for instance, background). And it's about 10 different colors in each file. Can we just add these instructions into files? Or there is another way?
btw, Now it just takes the missing instructions from special file and colorize ones into black. It works, but not as good as I want.

**Comment by optizone8** [comment]
Ok, already done. Last little thing. I can't find which instruction controls background color of current graph node
image

**Comment by xarkes** [comment]
Check here: https://github.com/radareorg/cutter/blob/master/src/utils/Configuration.cpp#L127

**Comment by optizone8** [comment]
Looks like 
image

**Comment by xarkes** [comment]
There is an error on OS X with qmake:
```

[1m../src/utils/ColorSchemeFileSaver.cpp:28:15: [0m[0;1;31merror: [0m[1mexpected value in expression[0m

#elif Q_OS_OSX
```
Can you fix it?

**Comment by optizone8** [comment]
I'll try

**Comment by fcasal** [comment]
Thanks for the contribution! I tried it on a linux system and there are some issues with the directories obtained by the ColorSchemeFileSaver constructor which do not correspond to the real ones. Also, the copy feature ColorSchemeFileSaver::copy quits early since the .template file does not exist and is not created since the open is in ReadOnly mode (http://doc.qt.io/qt-5/qfile.html#open).

**Comment by xarkes** [comment]
"Local data" (`%APPDATA%` for Windows, `~/.local/share` for Linux) is created during execution.
What I suggest, if everyone agrees, is that the theme loading should:
1. Check for the Cutter custom file, and load it if it exists
2. Use r2 console file (just call `ec theme`)

I don't think your `.template` and `.info` files are needed, I think you can get rid of that through `ecc`.

File: src/utils/ColorSchemeFileSaver.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+#include <QStandardPaths>
+#include "utils/Configuration.h"
+
+ColorSchemeFileSaver::ColorSchemeFileSaver(QObject* parent) : QObject (parent)
+{
+    QDir currDir;
+    QStringList dirs = QStandardPaths::standardLocations(QStandardPaths::GenericDataLocation);
+    dirs.removeOne(QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation));
+    currDir = QDir(dirs.at(0));
+    currDir.cd(currDir.absolutePath() + "/radare2");
```
Comment (xarkes) [inline src/utils/ColorSchemeFileSaver.cpp]: I would suggest to use `currentDir = QDir(dirs.at(0)).filePath("radare2");` which is shorter and cleaner.

File: src/utils/ColorSchemeFileSaver.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    // currDir.entryList(QDir::Dirs, QDir::Name) returns { current dir, upper dir, dirs ... }
+    // so it takes first (and only) dir using .at(2)
+    currDir.cd(currDir.absolutePath() + "/" + currDir.entryList(QDir::Dirs, QDir::Name).at(2) + "/");
```
Comment (xarkes) [inline src/utils/ColorSchemeFileSaver.cpp]: Same here, if you really need to use `+`, don't hardcode the slash, better use `QDir::separator()` which is cross platform.

File: src/utils/ColorSchemeFileSaver.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    standardR2ThemesLocationPath = currDir.absolutePath();
+
+    customR2ThemesLocationPath = QStandardPaths::writableLocation(QStandardPaths::GenericDataLocation) + "/Cutter/cons";
+}
+
+QFile::FileError ColorSchemeFileSaver::copy(const QString& srcThemeName, const QString& copyThemeName) const
+{
+    QFile templ(customR2ThemesLocationPath + "/.template");
+    QFile fIn(standardR2ThemesLocationPath + "/" + srcThemeName);
+    QFile fOut(customR2ThemesLocationPath + "/" + copyThemeName);
```
Comment (xarkes) [inline src/utils/ColorSchemeFileSaver.cpp]: Same here, use separator.

File: src/utils/ColorSchemeFileSaver.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    if (!templ.open(QFile::ReadOnly))
+        return templ.error();
+
```
Comment (xarkes) [inline src/utils/ColorSchemeFileSaver.cpp]: Add braces around the if.

File: src/utils/ColorSchemeFileSaver.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    for (auto &it : QDir(customR2ThemesLocationPath).entryInfoList())
+        if (it.fileName() == schemeName)
+            return true;
+    return false;
+}
+
+bool ColorSchemeFileSaver::isNameEngaged(const QString& name) const
+{
+    return QFile::exists(standardR2ThemesLocationPath + "/" + name) ||
+            QFile::exists(customR2ThemesLocationPath + "/" + name);
```
Comment (xarkes) [inline src/utils/ColorSchemeFileSaver.cpp]: `QDir::separator()`

File: src/utils/ColorSchemeFileSaver.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+QStringList ColorSchemeFileSaver::getSchemeCommands(const QString& schemeName) const
+{
+    QFile f(customR2ThemesLocationPath + "/" + schemeName);
+    if (!f.open(QFile::ReadOnly)) {
+        f.setFileName(standardR2ThemesLocationPath + "/" + schemeName);
```
Comment (xarkes) [inline src/utils/ColorSchemeFileSaver.cpp]: `QDir::separator()`

File: src/utils/Configuration.h [hunk 8]
```diff
@@ -8,6 +8,10 @@
 #define Config() (Configuration::instance())
 #define ConfigColor(x) Config()->getColor(x)
 
+constexpr const char *simpleTextOptionName = "gui.cflow";
+constexpr const char *defaultBackgroundOptionName = "gui.background";
+constexpr const char *path = "/home/optizone/git/cutter/src/themes/schemes/ColorScheme.json";
```
Comment (xarkes) [inline src/utils/Configuration.h]: Remove hardcoded paths. See @thestr4ng3r suggestions in the above comments.

**Comment by optizone8** [comment]
Which format of custom scheme files should I use:
1) "ec option rgb:hex_code_of_color"
2) "option hex_code_of_color"

And what do you mean saying 

> Use r2 console file (just call `ec theme`)

?
Because it already do this 
`for (auto &it : ColorSchemeFileWorker().getSchemeCommands(theme)) {`
`Core()->cmd(it);`
or did I misunderstand you?

**Comment by optizone8** [comment]
And about `.template` and `.info` files. 
Yes, `.template` is not necessary, because all information needed we can reach through `ecc` or `ecj`, but `.info` contains description of each option, so we can display it.

**Comment by xarkes** [comment]
> Which format of custom scheme files should I use:
    1. "ec option rgb:hex_code_of_color"
    2. "option hex_code_of_color"

For the custom Cutter theme? Having a json file looks good to me because it's easy to parse from within the application. Something like `{"btext": "000000"}` I guess.

> or did I misunderstand you?

I meant `eco` to load the theme, sorry. Like here: https://github.com/radareorg/cutter/blob/master/src/utils/Configuration.cpp#L262

> .info contains description of each option, so we can display it.

Then I think I didn't understand the purpose of it?

**Comment by optizone8** [comment]
> Then I think I didn't understand the purpose of it?
https://user-images.githubusercontent.com/42874998/45895407-a4241a00-bdc0-11e8-82b7-8847da2be052.png

Displays information about option near listView, under the color choose button (on the picture it is "info").

**Comment by xarkes** [comment]
Ah right, you mean the purpose of the color? Ok.

File: src/widgets/ColorSchemePrefWidget.cpp [hunk 9]
```diff
@@ -14,6 +14,87 @@
 #include <QColorDialog>
 #include <QJsonDocument>
 
+static const QMap<QString, QString> displayingTextMap = {
+        { "comment", QObject::tr("comment") },
```
Comment (xarkes) [inline src/widgets/ColorSchemePrefWidget.cpp]: `tr` is used for translations.
There is no need to use it here.

**Comment by ITAYC0HEN** [comment]
I checked out your branch and it seems like a really nice solution. Thank you :)

I think I have some bug where the color isn't shown at the box (circled):
image


Also, not that related to your PR, but I think "General" isn't a good name of this Preferences category. Seems like all there is there is Interface related settings.

**Comment by optizone8** [comment]
@ITAYC0HEN, I think that's OSX specific bug, because on my machine all works fine. I'll install OSX on my VM and find out what the problem is today.

**Comment by ITAYC0HEN** [comment]
Thank you! And it's Arch Linux 
You can easily download a ready-made VM here: osboxes.org/arch-linux/

**Comment by optizone8** [comment]
My machine is on Arch Linux too

File: src/widgets/ColorSchemePrefWidget.cpp (hunk 9 above)
Comment (optizone8) [reply src/widgets/ColorSchemePrefWidget.cpp]: This map was created to return displaying text, so `tr` is useful.

**Comment by optizone8** [comment]
Is there any description of each option? Because I can't add information and correct names to the `optionInfoMap`

**Comment by xarkes** [comment]
No I don't think there are any. I will ask around.

**Comment by xarkes** [comment]
Here is one I just made:
```
comment: r2 comments
usrcmt: user comments
args: function arguments
fname: function name
floc: function location
fline:
flag: flags color
label: label color
help: help color
flow:
flow2:
prompt: color of r2 prompt
offset: color of offsets
input:
invalid: invalid opcode color
other: other things color
b0x00:
b0x7f:
b0xff:
math: arithmetic color
bin:
btext:
push: push color
pop: pop color
crypto: cryptographic colors
jmp: jmp instruction color
cjmp: same?
call: call color
nop: nop color
ret: ret color
trap: trap color (interrupts)
swi: swi color
cmp: cmp color
reg: register color
creg: ?
num: immediate value color
mov: mov color
func_var: function variable name color
func_var_type: func variable (local or argument) type color
func_var_addr: func variable address color
widget_bg: ?
widget_sel: ?
ai.read:
ai.write:
ai.exec:
ai.seq:
ai.ascii:
graph.box: ?
graph.box2: ?
graph.box3: ?
graph.box4: ?
graph.true: jump arrow true
graph.false: jump arrow false
graph.trufae: jump arrow always
graph.cufrent: ?
graph.traced: ?
gui.cflow: ?
gui.dataoffset: ?
gui.background: background color
gui.alt_background: alternate background color
gui.border: box border color
wordhl: word highlight color
linehl: line highlight color
```

Not every color is used in Cutter afaik. Hope this helps.

**Comment by xarkes** [comment]
Can you rebase to fix conflicts? Thanks.

**Comment by xarkes** [comment]
I think there is an issue with your rebase... Please be careful and do not lose your work.

**Comment by optizone8** [comment]
So, I did something wrong with rebase?

**Comment by xarkes** [comment]
Yes, as you can see you are trying to push 65 commits that are not yours...
I don't know what you did, but you really messed up your branch and solving it with git won't be an easy task (especially if you came to that point).

What I suggest you to do, is to create a new branch on top of master (that means do `git checkout master; git pull;` make sure you are up to date with master `git reset --hard origin/master` and be careful with that command, it will erase any file and make it the same as on master, so make *sure* you are on the *master* branch when doing it)

After that just create your new clean branch: `git checkout -b mycolorupdate`
And now copy paste the new files and modification you made in that branch.
After that `git commit` and then `git push -f optizone8/color_scheme_feat` so this will erase everything that is in this PR with your beautiful new shiny commit.

I hope it's clear enough.

**Comment by xarkes** [comment]
Can you rebase now?
```
git checkout color_scheme_feat
git rebase master
...
# Fix conflicts
...
git rebase --continue
git push -f
```

**Comment by optizone8** [comment]
I did exactly the same last time, but before I rebased radare master onto my master. It didn't turn well. Next time i merged radare master into my master, same thing. But I need fresh master in my repo, so I can fix conflicts.
What should I do?

**Comment by xarkes** [comment]
Cool I will merge after next release.

**Comment by xarkes** [comment]
Indeed.

**Pull request merged**